firebase-appdistribution/src/test/java/com/google/firebase/appdistribution/ApkUpdaterTest.java

```diff
@@ -170,6 +183,52 @@ public void updateApk_whenInstallSuccessful_setsResult() throws Exception {
     assertThat(updateTask.isSuccessful()).isTrue();
   }
 
+  @Test
+  public void updateApk_whenSuccessfullyUpdated_notificationsSetCorrectly()
+      throws FirebaseAppDistributionException, ExecutionException, InterruptedException,
+          IOException {
+    doReturn(new ByteArrayInputStream(TEST_FILE.getBytes()))
+        .when(mockHttpsUrlConnection)
+        .getInputStream();
+    doNothing().when(apkUpdater).validateJarFile(any(), anyLong(), anyBoolean(), anyLong());
+    when(mockApkInstaller.installApk(any(), any())).thenReturn(Tasks.forResult(null));
+
+    UpdateTask updateTask = apkUpdater.updateApk(TEST_RELEASE, true);
+    updateTask.addOnCompleteListener(testExecutor, onCompleteListener);
+    List<UpdateProgress> events = new ArrayList<>();
+    updateTask.addOnProgressListener(testExecutor, events::add);
+    onCompleteListener.await();
+
+    assertThat(events).hasSize(3);
+    assertThat(events.get(0).getUpdateStatus()).isEqualTo(UpdateStatus.PENDING);
+    assertThat(events.get(1).getUpdateStatus()).isEqualTo(UpdateStatus.DOWNLOADING);
+    assertThat(events.get(1).getApkBytesDownloaded()).isEqualTo(TEST_FILE.length());
+    assertThat(events.get(2).getUpdateStatus()).isEqualTo(UpdateStatus.DOWNLOADED);
+  }
+
+  @Test
+  public void updateApk_invalidJarFile_throwsException() throws IOException {
```

Comment:
This test is named `invalidJarFile` but seems to actually test when the file input (or output?) fails.

Reply:
It tests that the tests fails when there is an invalidJarFile. My understanding was methodUnderTest_conditions_expectedEndState. So here the condition would be calling this with an invalidJarFile.

Reply:
But it looks like the condition is really more like "the file doesn't exist" (either when reading or writing, I'm not sure). From below: 
```
FileInputStream fileInputStream =
        (FileInputStream) ApkUpdaterTest.class.getResourceAsStream("file does not exist");
```

Reply:
There is no input or output happening with that fileInputStream.  It is simply so that this line " File apkFile = context.getFileStreamPath(fileName);" Returns an incorrect jarFile so that when we do " new JarFile(apkFile).close();" it throws an error. 

Either way the file not existing leads to an invalid jar file.

Reply:
Just to clarify the reason this fails is because we lack the validateJarFile mock. We don't pass in an invalid string by default the test will pass in an invalid string when it does. "context.getFileStreamPath(fileName);"